File: backend/src/main/java/com/erp/marks/MarksService.java
package com.erp.marks;

import com.erp.marks.dto.MarksRequest;

import java.util.List;
import java.util.Map;

public interface MarksService {
    Marks uploadMarks(MarksRequest request);
    List<Marks> uploadBulkMarks(List<MarksRequest> requests);
    Marks updateMarks(Long id, MarksRequest request);
    List<Marks> getMarksByStudent(Long studentId);
    List<Marks> getMarksByStudentAndSubject(Long studentId, Long subjectId);
    List<Marks> getMarksBySubject(Long subjectId);
    List<Marks> getMarksByExam(Long examId);
    Map<String, Object> generateStudentReport(Long studentId);
    Map<String, Object> generateSubjectReport(Long subjectId);
    Map<String, Object> getSubjectStatistics(Long subjectId);
    List<Map<String, Object>> getToppers(Long courseId, Integer semester);
}

File: frontend/src/services/marksService.js
import api from './api';

const marksService = {
    getMarksByStudent: async (studentId) => {
        const response = await api.get(`/marks/student/${studentId}`);
        return response.data;
    },
    getMarksBySubject: async (subjectId) => {
        const response = await api.get(`/marks/subject/${subjectId}`);
        return response.data;
    },
    getMarksByExam: async (examId) => {
        const response = await api.get(`/marks/exam/${examId}`);
        return response.data;
    },
    uploadMarks: async (marksData) => {
        const response = await api.post('/marks/upload', marksData);
        return response.data;
    },
    uploadBulkMarks: async (marksList) => {
        const response = await api.post('/marks/bulk', marksList);
        return response.data;
    },
    updateMarks: async (id, marksData) => {
        const response = await api.put(`/marks/${id}`, marksData);
        return response.data;
    },
    getStudentReport: async (studentId) => {
        const response = await api.get(`/marks/report/student/${studentId}`);
        return response.data;
    },
    getSubjectReport: async (subjectId) => {
        const response = await api.get(`/marks/report/subject/${subjectId}`);
        return response.data;
    },
    getSubjectStatistics: async (subjectId) => {
        const response = await api.get(`/marks/statistics/${subjectId}`);
        return response.data;
    },
    getToppers: async (courseId, semester) => {
        const response = await api.get(`/marks/toppers/${courseId}?semester=${semester}`);
        return response.data;
    }
};

export default marksService;
